File: tests/pages/LoginPage.ts
import type { Page, Locator, BrowserContext } from 'playwright';
import { ChatPage } from './ChatPage';
import { waitForLocatorVisible } from '../libs/waits';

export class LoginPage {
  readonly page: Page;
  readonly roomInput: Locator;
  readonly nameInput: Locator;
  readonly loginButton: Locator;
  readonly fandbInputs: Locator;
  readonly fandbForm: Locator;
  readonly hotelName: Locator;
  readonly notyfAnnouncer: Locator;

  constructor(page: Page) {
    this.page = page;
    this.roomInput = page.locator('#guest_room');
    this.nameInput = page.locator('#guest_name');
    this.loginButton = page.locator('#btn_login');
    this.fandbInputs = page.locator('.fandb-form-control-login');
    this.fandbForm = page.locator('.fandb-form');
    this.hotelName = page.locator('.hotel-name').first();
    // notyf may render notifications using several container/class names
    // depending on version or app markup. Match common variants and take
    // the first visible toast/container.
    this.notyfAnnouncer = page.locator('.notyf__toast, .notyf-announcer, .notyf');
    
  }

  async goto(url: string) {
    await this.page.goto(url);
  }

  async loginWithRoomAndName(room: string, name: string) {
    // Delegate to the unified `login` helper for the standard flow so
    // all login entry points funnel through a single method.
  }

  /**
   * General-purpose login helper that unifies the different login flows.
   * - `useFandb`: whether to use the F&B inputs variant (keeps existing assertions).
   * - `assertRoom`: when true, waits for the room heading after login.
   * This is a non-breaking convenience wrapper so tests and other helpers
   * can call a single method for most login cases.
   */
  async login(
    options: { room?: string; name?: string; mode?: 'fandb' | 'standard'; assertRoom?: boolean } = {}
  ) {
    const { room = '', name = '', mode = 'standard', assertRoom = false } = options;
    if (mode === 'fandb') {
      await this.loginWithFandbInputs(room, name, assertRoom);
      return;
    }
    await this.fillRoom(room);
    await this.fillName(name);
    await this.submitLogin(room || undefined, assertRoom);
  }

  // Minimal POM helpers: fill inputs and submit. Backward-compatible with existing helpers.
  async fillRoom(room: string) {
    await this.roomInput.fill(room);
  }

  async fillName(name: string) {
    await this.nameInput.fill(name);
  }

  /**
   * Submit the login form. Optionally assert the room heading appears when `assertRoom` is true
   * and `room` is provided.
   */
  async submitLogin(room?: string, assertRoom = false) {
    await waitForLocatorVisible(this.loginButton);
    await this.loginButton.click();
    // If the caller requested `assertRoom`, the subsequent `waitForRoom`
    // will cover waiting for the UI. Only perform the explicit load
    // wait when the caller does not request that assertion to avoid
    // redundant waiting.
    if (!assertRoom) {
      await this.page.waitForLoadState('domcontentloaded');
    }
    if (assertRoom && room) {
      await this.waitForRoom(room);
    }
  }

  async loginWithFandbInputs(room: string, name: string, assertRoom = false) {
    // Backwards-compatible wrapper — delegate to the centralized F&B
    // implementation to keep behavior consistent.
    return this._doFandbLogin(room, name, assertRoom);
  }

  // Centralized F&B login implementation used by both `login` (mode='fandb')
  // and the backward-compatible `loginWithFandbInputs` wrapper.
  private async _doFandbLogin(room: string, name: string, assertRoom = false) {
    await this.fandbInputs.first().waitFor({ state: 'visible' });
    await this.fandbInputs.last().waitFor({ state: 'visible' });
    if (await this.fandbInputs.count() !== 2) throw new Error('expected 2 F&B inputs');
    await this.fandbInputs.first().fill(room);                                         // nth(0)
    if ((await this.fandbInputs.first().getAttribute('id')) !== 'guest_room') throw new Error('expected first F&B input to have id guest_room');
    if ((await this.fandbInputs.first().inputValue()) !== room) throw new Error('expected room value to match');
    if ((await this.page.locator('#guest_room').inputValue()) !== room) throw new Error('expected #guest_room value to match');
    await this.fandbInputs.last().fill(name);                                          // nth(1)
    if ((await this.fandbInputs.last().getAttribute('id')) !== 'guest_name') throw new Error('expected last F&B input to have id guest_name');
    if ((await this.fandbInputs.last().inputValue()) !== name) throw new Error('expected guest_name value to match');
    if ((await this.page.locator('#guest_name').inputValue()) !== name) throw new Error('expected #guest_name value to match');
    await this.loginButton.click();
    // When `assertRoom` is true the following `waitForRoom` covers
    // waiting for the F&B form to appear; skip the explicit load
    // state wait in that case to reduce redundant waiting.
    if (!assertRoom) {
      await this.page.waitForLoadState('domcontentloaded');
    }
    // Only assert the room heading when caller requests it. Some tests
    // intentionally trigger failed logins and should not expect the room
    // heading to appear.
    if (assertRoom) {
      await this.waitForRoom(room);
    }
  }

  /**
   * Waits for the room heading inside the F&B form to be visible and returns its locator.
   * This centralizes the scoped lookup and timeout used by tests to avoid strict-mode
   * multiple-match errors when the same room appears in multiple page regions.
   */
  async waitForRoom(room: string, timeout = 15000) {
    const roomHeading = this.fandbForm.locator('h4.client-room', { hasText: room }).first();
    await roomHeading.waitFor({ state: 'visible', timeout });
    return roomHeading;
  }

  /**
   * Backwards-compatible alias used by tests: wait for the F&B form to display the
   * provided room. This centralizes any future selector changes in one place.
   */
  async waitForFandbToShow(room: string, timeout = 15000) {
    return this.waitForRoom(room, timeout);
  }

  /**
   * Clicks the chat entry point on the page and returns a `ChatPage` for the
   * newly opened chat window/tab. Centralizes the click selector and the
   * context.waitForEvent pairing to avoid duplicating the locator in tests.
   */
  async openChat(context: BrowserContext) {
    const [chatPage] = await Promise.all([
      context.waitForEvent('page'),
      this.page.locator('div.gradient h2.title', { hasText: 'Hablamos?' }).click(),
    ]);
    await chatPage.waitForLoadState('domcontentloaded');
    return new ChatPage(chatPage);
  }

  /**
   * High-level assertion helper verifying that the UI shows the provided
   * room and optional guest name. By default it asserts the F&B (fandb)
   * view, but callers can pass `mode: 'standard'` to check the global
   * post-login UI selectors instead.
   *
   * Backwards compatible: previous calls with (room, name, timeout)
   * continue to work because `mode` is the optional fourth parameter.
   */
  async assertLoggedIn(
    room: string,
    name?: string,
    timeout = 15000,
    mode: 'fandb' | 'standard' = 'fandb'
  ) {
    if (mode === 'fandb') {
      await this.waitForRoom(room, timeout);
      await this.hotelName.waitFor({ state: 'visible', timeout });
      if (name) {
        const text = await this.fandbForm.textContent();
        if (!text || !text.includes(name)) throw new Error(`expected F&B form to contain ${name}`);
      }
      {
        const text = await this.fandbForm.textContent();
        if (!text || !text.includes(room)) throw new Error(`expected F&B form to contain ${room}`);
      }
      return;
    }

    // Standard view: use global locators instead of the F&B scoped form.
    const roomLocator = this.page.locator('h4.client-room', { hasText: room }).first();
    const nameLocator = this.page.locator('h4.client-name', { hasText: name || '' }).first();
    await roomLocator.waitFor({ state: 'visible', timeout });
    await this.hotelName.waitFor({ state: 'visible', timeout });
    if (name) {
      await nameLocator.waitFor({ state: 'visible', timeout });
    }
  }

  // --- Notification helpers (notyf) ---
  /**
   * Wait for the global notyf announcer to become visible and return its locator.
   */
  async waitForNotification(timeout = 5000) {
    // Wait for any common notyf selector to become visible in the DOM.
    await this.page.waitForSelector('.notyf__toast, .notyf-announcer, .notyf, .notyf__toast-message', { state: 'visible', timeout });
    // Return a locator that resolves to the first visible toast/message container.
    const visible = this.page.locator('.notyf__toast:visible, .notyf-announcer:visible, .notyf:visible, .notyf__toast-message:visible');
    return visible.first();
  }

  /**
   * Assert that the notyf announcer contains the provided text or regex.
   */
  async assertNotificationContains(expected: string | RegExp, timeout = 5000) {
    // If we have a concrete expected value, wait for a toast that contains it.
    const sel = '.notyf__toast, .notyf-announcer, .notyf, .notyf__toast-message';
    try {
      const byText = this.page.locator(sel, { hasText: expected });
      await byText.first().waitFor({ state: 'visible', timeout });
      const txt = await byText.first().innerText();
      if (!txt) throw new Error('notification empty');
      if (typeof expected === 'string') {
        if (!txt.includes(expected)) throw new Error(`expected notification to contain ${expected}`);
      } else {
        if (!expected.test(txt)) throw new Error(`expected notification to match ${expected}`);
      }
      return;
    } catch (err) {
      // Fallback: wait for any visible toast and read its text for diagnostic failure message.
      const locator = await this.waitForNotification(timeout);
      const txt = await locator.innerText().catch(() => null);
      if (!txt) throw new Error('notification empty (fallback)');
      if (typeof expected === 'string') {
        if (!txt.includes(expected)) throw new Error(`expected notification to contain ${expected}; actual: ${txt}`);
      } else {
        if (!expected.test(txt)) throw new Error(`expected notification to match ${expected}; actual: ${txt}`);
      }
    }
  }

  /**
   * Assert the notyf announcer is visible.
   */
  async assertNotificationVisible(timeout = 5000) {
    await this.page.waitForSelector('.notyf__toast, .notyf-announcer, .notyf', { state: 'visible', timeout });
  }

  /**
   * Assert that the notyf announcer contains all provided strings/regexes.
   * Waits once for visibility, then checks each expected value.
   */
  async assertNotificationContainsAll(expected: Array<string | RegExp>, timeout = 5000) {
    const locator = await this.waitForNotification(timeout);
    const txt = await locator.innerText().catch(() => null);
    if (!txt) throw new Error('notification empty');
    for (const e of expected) {
      if (typeof e === 'string') {
        if (!txt.includes(e)) throw new Error(`expected notification to contain ${e}`);
      } else {
        if (!e.test(txt)) throw new Error(`expected notification to match ${e}`);
      }
    }
  }
}

File: tests/pages_py/login_page.py
import re
from typing import Optional, Literal
from playwright.sync_api import Page, BrowserContext
from .chat_page import ChatPage


class LoginPage:
    def __init__(self, page: Page):
        self.page = page
        self.room_input = page.locator('#guest_room')
        self.name_input = page.locator('#guest_name')
        self.login_button = page.locator('#btn_login')
        self.fandb_inputs = page.locator('.fandb-form-control-login')
        self.fandb_form = page.locator('.fandb-form')
        self.hotel_name = page.locator('.hotel-name').first
        # notyf may render notifications using several container/class names
        self.notyf_announcer = page.locator('.notyf__toast, .notyf__toast-message, .notyf-announcer, .notyf').first

    def goto(self, url: str):
        self.page.goto(url)

    def login(
        self,
        room: str = '',
        name: str = '',
        mode: Literal['fandb', 'standard'] = 'standard',
        assert_room: bool = False
    ):
        """General-purpose login helper that unifies different login flows."""
        if mode == 'fandb':
            self._login_with_fandb_inputs(room, name, assert_room)
            return
        
        self.fill_room(room)
        self.fill_name(name)
        self.submit_login(room if room else None, assert_room)

    def fill_room(self, room: str):
        self.room_input.fill(room)

    def fill_name(self, name: str):
        self.name_input.fill(name)

    def submit_login(self, room: Optional[str] = None, assert_room: bool = False):
        """Submit the login form."""
        self.login_button.wait_for(state='visible')
        self.login_button.click()
        
        if not assert_room:
            self.page.wait_for_load_state('domcontentloaded')
        
        if assert_room and room:
            self.wait_for_room(room)

    def _login_with_fandb_inputs(self, room: str, name: str, assert_room: bool = False):
        """Centralized F&B login implementation."""
        self.fandb_inputs.first.wait_for(state='visible')
        self.fandb_inputs.last.wait_for(state='visible')
        
        if self.fandb_inputs.count() != 2:
            raise ValueError('expected 2 F&B inputs')
        
        # Fill room (first input)
        self.fandb_inputs.first.fill(room)
        if self.fandb_inputs.first.get_attribute('id') != 'guest_room':
            raise ValueError('expected first F&B input to have id guest_room')
        if self.fandb_inputs.first.input_value() != room:
            raise ValueError('expected room value to match')
        
        # Fill name (last input)
        self.fandb_inputs.last.fill(name)
        if self.fandb_inputs.last.get_attribute('id') != 'guest_name':
            raise ValueError('expected last F&B input to have id guest_name')
        if self.fandb_inputs.last.input_value() != name:
            raise ValueError('expected guest_name value to match')
        
        self.login_button.click()
        
        if not assert_room:
            self.page.wait_for_load_state('domcontentloaded')
        
        if assert_room:
            self.wait_for_room(room)

    def wait_for_room(self, room: str, timeout: int = 15000):
        """Waits for the room heading inside the F&B form to be visible."""
        room_heading = self.fandb_form.locator('h4.client-room', has_text=room).first
        room_heading.wait_for(state='visible', timeout=timeout)
        return room_heading

    def open_chat(self, context: BrowserContext) -> ChatPage:
        """Clicks the chat entry point and returns a ChatPage for the newly opened window."""
        with context.expect_page() as new_page_info:
            self.page.locator('div.gradient h2.title', has_text='Hablamos?').click()
        
        chat_page = new_page_info.value
        chat_page.wait_for_load_state('domcontentloaded')
        return ChatPage(chat_page)

    def assert_logged_in(
        self,
        room: str,
        name: Optional[str] = None,
        timeout: int = 15000,
        mode: Literal['fandb', 'standard'] = 'fandb'
    ):
        """High-level assertion verifying the UI shows the provided room and guest name."""
        if mode == 'fandb':
            self.wait_for_room(room, timeout)
            self.hotel_name.wait_for(state='visible', timeout=timeout)
            
            if name:
                text = self.fandb_form.text_content()
                if not text or name not in text:
                    raise AssertionError(f"expected F&B form to contain {name}")
            
            text = self.fandb_form.text_content()
            if not text or room not in text:
                raise AssertionError(f"expected F&B form to contain {room}")
            return

        # Standard view: use global locators
        room_locator = self.page.locator('h4.client-room', has_text=room).first
        room_locator.wait_for(state='visible', timeout=timeout)
        self.hotel_name.wait_for(state='visible', timeout=timeout)
        
        if name:
            name_locator = self.page.locator('h4.client-name', has_text=name).first
            name_locator.wait_for(state='visible', timeout=timeout)

    def wait_for_notification(self, timeout: int = 5000):
        """Wait for the global notyf announcer to become visible."""
        preferred = [
            '.notyf__toast:visible',
            '.notyf__toast-message:visible',
            '.notyf-announcer:visible',
            '.notyf:visible'
        ]
        
        for sel in preferred:
            if self.page.locator(sel).count() > 0:
                return self.page.locator(sel).first
        
        # Fallback
        self.page.wait_for_selector(
            '.notyf__toast, .notyf-announcer, .notyf, .notyf__toast-message',
            state='visible',
            timeout=timeout
        )
        return self.page.locator('.notyf__toast:visible, .notyf-announcer:visible, .notyf:visible, .notyf__toast-message:visible').first

    def assert_notification_contains(self, expected: str | re.Pattern, timeout: int = 5000):
        """Assert that the notyf announcer contains the provided text or regex."""
        sel = '.notyf__toast, .notyf-announcer, .notyf, .notyf__toast-message'
        
        try:
            by_text = self.page.locator(sel, has_text=expected)
            by_text.first.wait_for(state='visible', timeout=timeout)
            txt = by_text.first.inner_text()
            
            if not txt:
                raise ValueError('notification empty')
            
            if isinstance(expected, str):
                if expected not in txt:
                    raise AssertionError(f"expected notification to contain {expected}")
            else:
                if not expected.search(txt):
                    raise AssertionError(f"expected notification to match {expected}")
            return
        except Exception:
            # Fallback: aggregate text from all matching elements
            locator = self.wait_for_notification(timeout)
            try:
                txt = locator.inner_text()
            except Exception:
                txt = ''
            
            if not txt:
                txt = self.page.evaluate(f'''(s) => {{
                    const nodes = Array.from(document.querySelectorAll(s));
                    return nodes.map(n => (n.textContent || '').trim()).filter(Boolean).join(' ').trim();
                }}''', sel)
            
            if not txt:
                raise ValueError('notification empty (fallback)')
            
            if isinstance(expected, str):
                if expected not in txt:
                    raise AssertionError(f"expected notification to contain {expected}; actual: {txt}")
            else:
                if not expected.search(txt):
                    raise AssertionError(f"expected notification to match {expected}; actual: {txt}")

    def assert_notification_visible(self, timeout: int = 5000):
        """Assert the notyf announcer is visible."""
        self.page.wait_for_selector(
            '.notyf__toast, .notyf-announcer, .notyf',
            state='visible',
            timeout=timeout
        )

    def assert_notification_contains_all(self, expected: list[str | re.Pattern], timeout: int = 5000):
        """Assert that the notyf announcer contains all provided strings/regexes."""
        locator = self.wait_for_notification(timeout)
        
        try:
            txt = locator.inner_text()
        except Exception:
            txt = ''
        
        if not txt:
            sel = '.notyf__toast, .notyf-announcer, .notyf__toast-message, .notyf'
            txt = self.page.evaluate(f'''(s) => {{
                const nodes = Array.from(document.querySelectorAll(s));
                return nodes.map(n => (n.textContent || '').trim()).filter(Boolean).join(' ').trim();
            }}''', sel)
        
        if not txt:
            txt = self.page.locator('body').inner_text()
        
        if not txt:
            raise ValueError('notification empty')
        
        # Normalize whitespace
        normalized = re.sub(r'\s+', ' ', txt).strip()
        
        for e in expected:
            try:
                locator = self.page.locator('body', has_text=e)
                locator.first.wait_for(state='visible', timeout=timeout)
                continue
            except Exception:
                pass
            
            if isinstance(e, str):
                if e not in normalized:
                    raise AssertionError(f"expected notification to contain {e}; actual: {normalized}")
            else:
                if not e.search(normalized):
                    raise AssertionError(f"expected notification to match {e}; actual: {normalized}")
